File: medium/107. Binary Tree Level Order Traversal II/107. Binary Tree Level Order Traversal II.py
from collections import deque
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrderBottom(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        res: List[List[int]] = []
        queue: deque[TreeNode] = deque([root])

        while len(queue) > 0:
            level_items: List[int] = []
            level_length = len(queue)

            for _ in range(0, level_length):
                curr_node = queue.popleft()

                level_items.append(curr_node.val)
                if curr_node.left:
                    queue.append(curr_node.left)
                if curr_node.right:
                    queue.append(curr_node.right)

            res.append(level_items)

        res.reverse()
        return res

File: medium/107. Binary Tree Level Order Traversal II/107. Binary Tree Level Order Traversal II.ts


class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

function levelOrder(root: TreeNode | null): number[][] {
    if (!root) return []

    const res: number[][] = []
    const queue: TreeNode[] = [root]

    while (queue.length > 0) {
        const levelItems: number[] = []
        const levelSize = queue.length

        for (let i = 0; i < levelSize; i++) {
            const currentNode = queue.shift()!

            levelItems.push(currentNode.val)
            if (currentNode.left) queue.push(currentNode.left)
            if (currentNode.right) queue.push(currentNode.right)
        }

        res.push(levelItems)
    }

    res.reverse()
    return res
};

export { }